@uranusjr Thanks, I am able to reproduce it locally. I am adding cachetools dependency. I do see `ignore_missing_imports=True` in pyproject.yaml but see types added in devel-common/pyproject.toml for mypy. It seems adding `types-cachetools` in devel-common seems to help.

https://github.com/apache/airflow/blob/fb21525f83cc759d359419df52a36996dcf88f09/pyproject.toml#L1004-L1006

https://github.com/apache/airflow/blob/fb21525f83cc759d359419df52a36996dcf88f09/devel-common/pyproject.toml#L107-L130

```
prek --color always --verbose --hook-stage manual "mypy-providers" --all-files
Running hooks for `providers`:
Run mypy for providers (manual)..........................................Failed
- hook id: mypy-providers
- duration: 282.93s
- exit code: 1

  Running mypy with @/files/mypy_files.txt
  You cand check the list of files in: /home/karthikeyan/stuff/python/airflow/files/mypy_files.txt
  Running mypy with @/files/mypy_files.txt
  Regenerating provider dependencies file
  Refreshed 97 providers with 1983 Python files.

  Written /home/karthikeyan/stuff/python/airflow/generated/provider_dependencies.json

  Docker container engine detected.
  Attempting to generate provider dependencies. 2 pyproject.toml file(s) changed 
  since last check.

  Using 'uv' to install Airflow

  Using airflow version from current sources

  providers/fab/src/airflow/providers/fab/auth_manager/fab_auth_manager.py:25: error:
  Library stubs not installed for "cachetools"  [import-untyped]
      from cachetools import TTLCache, cachedmethod
      ^
  providers/fab/src/airflow/providers/fab/auth_manager/fab_auth_manager.py:25: note: Hint: "python3 -m pip install types-cachetools"
  providers/fab/src/airflow/providers/fab/auth_manager/fab_auth_manager.py:25: note: (or run "mypy --install-types" to install all missing stub packages)
  providers/fab/src/airflow/providers/fab/auth_manager/fab_auth_manager.py:25: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
  Found 1 error in 1 file (checked 4075 source files)
  Error 1 returned
  If you see strange stacktraces above, and can't reproduce it, please run this command and try again:

  breeze ci-image build --python 3.10

  You can also run `breeze down --cleanup-mypy-cache` to clean up the cache used.
```


```
prek --color always --verbose --hook-stage manual "mypy-providers" --all-files
Running hooks for `providers`:
Run mypy for providers (manual)..........................................Passed
- hook id: mypy-providers
- duration: 261.46s

  Running mypy with @/files/mypy_files.txt
  You cand check the list of files in: /home/karthikeyan/stuff/python/airflow/files/mypy_files.txt
  Running mypy with @/files/mypy_files.txt
  Docker container engine detected.

  Using 'uv' to install Airflow

  Using airflow version from current sources

  Success: no issues found in 4075 source files
```

--- a/providers/fab/docs/index.rst
+++ b/providers/fab/docs/index.rst
@@ -121,7 +121,7 @@ PIP package                                 Version required
 ``jmespath``                                ``>=0.7.0; python_version < "3.13"``
 ``werkzeug``                                ``>=2.2,<4; python_version < "3.13"``
 ``wtforms``                                 ``>=3.0,<4; python_version < "3.13"``
-``cachetools``                              ``>=5.0; python_version < "3.13"``
+``cachetools``                              ``>=6.0; python_version < "3.13"``
 ``flask_limiter``                           ``>3,!=3.13,<4``
 ==========================================  ==========================================
 

--- a/providers/fab/docs/index.rst
+++ b/providers/fab/docs/index.rst
@@ -121,6 +121,7 @@ PIP package                                 Version required
 ``jmespath``                                ``>=0.7.0; python_version < "3.13"``
 ``werkzeug``                                ``>=2.2,<4; python_version < "3.13"``
 ``wtforms``                                 ``>=3.0,<4; python_version < "3.13"``
+``cachetools``                              ``>=5.0; python_version < "3.13"``
 ``flask_limiter``                           ``>3,!=3.13,<4``
 ==========================================  ==========================================
 

--- a/providers/fab/provider.yaml
+++ b/providers/fab/provider.yaml
@@ -250,6 +250,14 @@ config:
         type: integer
         example: ~
         default: "1"
+      cache_ttl:
+        description: |
+          Number of seconds after which the user cache will expire to refetch updated user and
+          permissions.
+        version_added: 3.1.2
+        type: integer
+        example: ~
+        default: "30"
 
 auth-managers:
   - airflow.providers.fab.auth_manager.fab_auth_manager.FabAuthManager

--- a/providers/fab/pyproject.toml
+++ b/providers/fab/pyproject.toml
@@ -81,7 +81,7 @@ dependencies = [
     "jmespath>=0.7.0; python_version < '3.13'",
     "werkzeug>=2.2,<4; python_version < '3.13'",
     "wtforms>=3.0,<4; python_version < '3.13'",
-    "cachetools>=5.0; python_version < '3.13'",
+    "cachetools>=6.0; python_version < '3.13'",
 
     # https://github.com/dpgaspar/Flask-AppBuilder/blob/release/4.6.3/setup.py#L54C8-L54C26
     # with an exclusion to account for https://github.com/alisaifee/flask-limiter/issues/479

--- a/providers/fab/pyproject.toml
+++ b/providers/fab/pyproject.toml
@@ -81,6 +81,7 @@ dependencies = [
     "jmespath>=0.7.0; python_version < '3.13'",
     "werkzeug>=2.2,<4; python_version < '3.13'",
     "wtforms>=3.0,<4; python_version < '3.13'",
+    "cachetools>=5.0; python_version < '3.13'",
 
     # https://github.com/dpgaspar/Flask-AppBuilder/blob/release/4.6.3/setup.py#L54C8-L54C26
     # with an exclusion to account for https://github.com/alisaifee/flask-limiter/issues/479

--- a/providers/fab/src/airflow/providers/fab/auth_manager/fab_auth_manager.py
+++ b/providers/fab/src/airflow/providers/fab/auth_manager/fab_auth_manager.py
@@ -22,7 +22,7 @@ from pathlib import Path
 from typing import TYPE_CHECKING, Any
 from urllib.parse import urljoin
 
-from cachetools import TTLCache, cached
+from cachetools import TTLCache, cachedmethod
 from connexion import FlaskApi
 from fastapi import FastAPI
 from fastapi.middleware.wsgi import WSGIMiddleware
@@ -162,7 +162,7 @@ _MAP_MENU_ITEM_TO_FAB_RESOURCE_TYPE = {
     MenuItem.XCOMS: RESOURCE_XCOM,
 }
 
-CACHE_TTL = conf.getint("fab", "cache_ttl", fallback=10)
+CACHE_TTL = conf.getint("fab", "cache_ttl", fallback=30)
 
 if AIRFLOW_V_3_1_PLUS:
     from airflow.providers.fab.www.security.permissions import RESOURCE_HITL_DETAIL
@@ -178,6 +178,7 @@ class FabAuthManager(BaseAuthManager[User]):
     This auth manager is responsible for providing a backward compatible user management experience to users.
     """
 
+    cache: TTLCache = TTLCache(maxsize=1024, ttl=CACHE_TTL)
     appbuilder: AirflowAppBuilder | None = None
 
     def init_flask_resources(self) -> None:
@@ -255,7 +256,7 @@ class FabAuthManager(BaseAuthManager[User]):
 
         return current_user
 
-    @cached(cache=TTLCache(maxsize=1024, ttl=CACHE_TTL), key=lambda _, token: int(token["sub"]))
+    @cachedmethod(lambda self: self.cache, key=lambda _, token: int(token["sub"]))
     def deserialize_user(self, token: dict[str, Any]) -> User:
         id = int(token["sub"])
 

--- a/providers/fab/src/airflow/providers/fab/auth_manager/fab_auth_manager.py
+++ b/providers/fab/src/airflow/providers/fab/auth_manager/fab_auth_manager.py
@@ -22,6 +22,7 @@ from pathlib import Path
 from typing import TYPE_CHECKING, Any
 from urllib.parse import urljoin
 
+from cachetools import TTLCache, cached
 from connexion import FlaskApi
 from fastapi import FastAPI
 from fastapi.middleware.wsgi import WSGIMiddleware
@@ -161,6 +162,7 @@ _MAP_MENU_ITEM_TO_FAB_RESOURCE_TYPE = {
     MenuItem.XCOMS: RESOURCE_XCOM,
 }
 
+CACHE_TTL = conf.getint("fab", "cache_ttl", fallback=10)
 
 if AIRFLOW_V_3_1_PLUS:
     from airflow.providers.fab.www.security.permissions import RESOURCE_HITL_DETAIL
@@ -176,7 +178,6 @@ class FabAuthManager(BaseAuthManager[User]):
     This auth manager is responsible for providing a backward compatible user management experience to users.
     """
 
-    user_cache: dict[int, User] = {}
     appbuilder: AirflowAppBuilder | None = None
 
     def init_flask_resources(self) -> None:
@@ -254,16 +255,12 @@ class FabAuthManager(BaseAuthManager[User]):
 
         return current_user
 
+    @cached(cache=TTLCache(maxsize=1024, ttl=CACHE_TTL), key=lambda _, token: int(token["sub"]))
     def deserialize_user(self, token: dict[str, Any]) -> User:
         id = int(token["sub"])
-        user = self.user_cache.get(id)
 
-        if not user:
-            with create_session() as session:
-                user = session.scalars(select(User).where(User.id == id)).one()
-                self.user_cache[id] = user
-
-        return user
+        with create_session() as session:
+            return session.scalars(select(User).where(User.id == id)).one()
 
     def serialize_user(self, user: User) -> dict[str, Any]:
         return {"sub": str(user.id)}

--- a/providers/fab/src/airflow/providers/fab/get_provider_info.py
+++ b/providers/fab/src/airflow/providers/fab/get_provider_info.py
@@ -177,6 +177,13 @@ def get_provider_info():
                         "example": None,
                         "default": "1",
                     },
+                    "cache_ttl": {
+                        "description": "Number of seconds after which the user cache will expire to refetch updated user and\npermissions.\n",
+                        "version_added": "3.1.2",
+                        "type": "integer",
+                        "example": None,
+                        "default": "30",
+                    },
                 },
             }
         },

--- a/providers/fab/tests/unit/fab/auth_manager/test_fab_auth_manager.py
+++ b/providers/fab/tests/unit/fab/auth_manager/test_fab_auth_manager.py
@@ -16,6 +16,7 @@
 # under the License.
 from __future__ import annotations
 
+import time
 from contextlib import contextmanager, suppress
 from itertools import chain
 from typing import TYPE_CHECKING
@@ -198,7 +199,9 @@ class TestFabAuthManager:
             with user_set(minimal_app_for_auth_api, flask_g_user):
                 assert auth_manager.get_user() == flask_g_user
 
+    @conf_vars({("fab", "cache_ttl"): "1"})
     def test_deserialize_user(self, flask_app, auth_manager_with_appbuilder):
+        """Test user objects are cached and that the cache expires after configured TTL."""
         user = create_user(flask_app, "test")
         with assert_queries_count(2):
             result = auth_manager_with_appbuilder.deserialize_user({"sub": str(user.id)})
@@ -206,6 +209,12 @@ class TestFabAuthManager:
         assert user.get_id() == result.get_id()
 
         with assert_queries_count(0):
+            result = auth_manager_with_appbuilder.deserialize_user({"sub": str(user.id)})
+
+        assert user.get_id() == result.get_id()
+
+        time.sleep(1)
+        with assert_queries_count(2):
             result = auth_manager_with_appbuilder.deserialize_user({"sub": str(user.id)})
 
         assert user.get_id() == result.get_id()

--- a/providers/fab/tests/unit/fab/auth_manager/test_fab_auth_manager.py
+++ b/providers/fab/tests/unit/fab/auth_manager/test_fab_auth_manager.py
@@ -200,12 +200,10 @@ class TestFabAuthManager:
 
     def test_deserialize_user(self, flask_app, auth_manager_with_appbuilder):
         user = create_user(flask_app, "test")
-        assert not auth_manager_with_appbuilder.user_cache
         with assert_queries_count(2):
             result = auth_manager_with_appbuilder.deserialize_user({"sub": str(user.id)})
 
         assert user.get_id() == result.get_id()
-        assert auth_manager_with_appbuilder.user_cache[user.id] == user
 
         with assert_queries_count(0):
             result = auth_manager_with_appbuilder.deserialize_user({"sub": str(user.id)})

